zorancutura_some_ip_gateway (#1115)
eclipse-score/score: docs/features/communication/some_ip_gateway/architecture/index.rst

     -
     - x
   * - Timeout
     - Detects if no message is received within a defined time window.
     -
     - x

E2E Design and Interface Considerations
---------------------------------------
As mentioned above, not all E2E checks can be kept hidden within the gateway. For some problems it is up to the application to decide
whether it is still valuable to access and process the data. This is in particular true for duplication or re-odering issues.
Therefore, it is required to pass a **tupel** consisting of the **payload data** as well as **supplementary E2E metadata and error information**
to the IPC clients to enable the client to individually judge on particular E2E issues.

* The API design needs to put the payload information as well as the additional E2E metadata as close as possible together to easily enable and motivate clients to consequently check for potential errors.
* Additional metadata and error information needs to be incorporated into regular mw::com user interface
* Error related interface design needs to be highly optimezed for the "good case" to have an optimized support for the common IPC case

> * AoU's need to be provided to force the user to check for the (E2E-) result

Author: done

* Additional E2E metadata to be handed over to the clients:

  * Counter, slightly different interpretation depending on profile, n/a in case the profile doesn't support it
  * Data ID, n/a in case the profile does not support it or if the Data ID is not explicitely transmitted like in profile 22
  * Profile identification, required to properly interpret E2E attributes like "Counter". (Could be either an "Alive Counter" or a "Sequence Counter" depending on the profile)
  * Detailed, profile specific error code (enum)

* Proposed Error Enumeration

  * No E2E error

>   * CRC Error

Reviewer: It might make sense to map CRC errors always to sequence errors, since we might have a corruption of service/instance id and would therefore report the error to the wrong recipient.

Reviewer: The problem is how do we deliver a CRC error if we do not know whether ServiceID or InstanceID were corrupted? They are also protected by the same CRC after all.
So worst case, we would report the CRC error to somebody completely unrelated. This could cause havoc in a communication that did not have any E2E failures in reality.

The only safe thing we can do in the gateway is to discard the message completely and not forward it at all.
This would result in a sequence error at the client once the next message comes in, or in a timeout error if all following messages fail the CRC check as well.

All in all this is not something we can decide now. I just wanted to raise it. Maybe you can add a note that CRC errors are problematic in the gateway because we cannot safely forward such messages to the correct recipients.

   * SYNC
   * INITIAL

E2E State Machine Considerations
--------------------------------
The E2E (End-to-End) state machine as defined within AUTOSAR E2E protocol provides a summarized result
about the overall health and state of a communication channel. Unlike individual E2E Profile Check() functions,
which assess data validity for a single communication cycle, the state machine aggregates results from multiple Check()
function invocations over a period. This allows it to determine a more holistic and debounced status of the communication.

Purpose of the E2E State Machine:
The primary purpose of the E2E state machine is to transform instantaneous "per-cycle" check results into a stable,
long-term communication channel status. This aggregated status is then provided to the consuming application,
enabling it to make informed decisions about whether the received data can be trusted and used for safety-related functions.

As mentioned above, the E2E statemachine is associated to one communication channel which is in turn associated to exactly one
individual IPC client. Therefore it is an obvious consequence, that the individiual state machine handling and state machine

> configuration is responsibility of the client and not a central responsibility of the gateway.

Author: done

The diagram below outlines this distribution of responsibilties.

*Diagram: E2E state machine responsibility associated to IPC client*

.. figure:: ../assets/e2e_state_machine_on_client_side.drawio.svg
   :align: center
   :name: _e2e_state_machine_on_client_side

   E2E state machine responsibility associated to IPC client

**Considered Alternative**

If we allocate the statemachine responsibility to the gateway the distribution of resposnibilities would look like in the following diagram

*Diagram: E2E state machine responsibility associated to the gateway*

.. figure:: ../assets/e2e_state_machine_in_gateway.drawio.svg
   :align: center
   :name: _e2e_state_machine_in_gateway

   E2E state machine responsibility associated to the gateway

> Due to pub/sub nature of mw::com, clients listening on the same topic can not be separately addressed. Therefore, **the state machine results

Reviewer 2: ```suggestion
Due to pub/sub nature of mw::com, clients listening on the same topic cannot be separately addressed. Therefore, **the state machine results
```
